### PR TITLE
Do not consume unmatched logs in LogCollector

### PR DESCRIPTION
**Pull request by keeferrourke**
Addresses #2128.

Open to other solutions! Maybe this is actually a bad idea -- my use-case may the outlier here, but I find the current behaviour very odd that `takeMessages(filter)` consumes the whole queue of collected logs and only returns the ones that matched.

### CODE REVIEW

File: wisp-logging-testing/src/main/kotlin/wisp/logging/WispQueuedLogCollector.kt [hunk 1]
```diff
@@ -37,38 +37,34 @@ class WispQueuedLogCollector : LogCollector {
...
+    queue.removeAll(resultList)
+    return resultList
   }
 
   override fun takeEvent(loggerClass: KClass<*>?, minLevel: Level, pattern: Regex?): ILoggingEvent {
-    while (true) {
-      val event = takeOrNull(loggerClass, minLevel, pattern)
-      if (event != null) return event
+    require(wasStarted) { "not collecting logs: did you forget to start the service?" }
+    for (i in 1..5) {
```
[the remark this answers is not in the document]
Comment (keeferrourke) [reply wisp-logging-testing/src/main/kotlin/wisp/logging/WispQueuedLogCollector.kt:48]: More closely matches (but not perfectly) the `poll(500)` in the previous impl.

File: misk-testing/src/test/kotlin/misk/logging/LogCollectorTest.kt [hunk 2]
```diff
@@ -14,8 +15,9 @@ import kotlin.test.assertFailsWith
 
 @MiskTest(startService = true)
 class LogCollectorTest {
```
Comment (keeferrourke) [inline misk-testing/src/test/kotlin/misk/logging/LogCollectorTest.kt:17]: I don't particularly enjoy that this test appears to be a copy of the wisp LogCollectorTest, but 🤷  I added the new test here too.

File: misk-testing/src/main/kotlin/misk/logging/RealLogCollector.kt [hunk 3]
```diff
@@ -10,37 +10,12 @@ import javax.inject.Singleton
...
 
 @Singleton
-internal class RealLogCollector @Inject constructor() :
-  AbstractIdleService(),
-  LogCollector,
+internal class RealLogCollector @Inject constructor(
+  private val wispQueuedLogCollector: WispQueuedLogCollector
+) : AbstractIdleService(),
+  LogCollector by wispQueuedLogCollector,
   LogCollectorService {
```
Comment (keeferrourke) [inline misk-testing/src/main/kotlin/misk/logging/RealLogCollector.kt:17]: I think ideally `LogCollectorService` could be delegated to the `WispQueuedLogCollector`, but I gues wisp-logging-testing doesn't want to depend on Guava?

File: wisp-logging-testing/src/main/kotlin/wisp/logging/LogCollector.kt [hunk 4]
```diff
@@ -43,7 +43,17 @@ interface LogCollector {
...
+
+  /**
+   * Takes all matching messages, optionally leaving unmatched logs in this collector.
+   */
+  fun takeMessages(
+    loggerClass: KClass<*>? = null,
+    minLevel: Level = Level.INFO,
+    pattern: Regex? = null,
+    consumeUnmatchedLogs: Boolean = false,
   ): List<String>
```
Comment (keeferrourke) [inline wisp-logging-testing/src/main/kotlin/wisp/logging/LogCollector.kt:57]: I think maybe this is the interface we want to land on eventually if some folks depend on consuming all unmatched logs?

Comment (r3mariano) [reply wisp-logging-testing/src/main/kotlin/wisp/logging/LogCollector.kt:57]: Yep, looks good.

**Review by r3mariano — APPROVED** [review]

**Pull request merged**